x-pack/plugin/core/src/main/java/module-info.java

```diff
@@ -210,4 +211,6 @@
     exports org.elasticsearch.xpack.core.watcher.trigger;
     exports org.elasticsearch.xpack.core.watcher.watch;
     exports org.elasticsearch.xpack.core.watcher;
+
+    opens org.elasticsearch.license.internal; // spi
```

Comment:
Mostly for my own learning: I understand above `exports` means only the stateless package can access the internal classes. For this `opens` statement, I think it means the internal classes can be accessed reflectively on runtime which is supposedly needed for SPI initialization. My questions are:
* Why we didn't have a similar opens statement for security extension? But it still works?
* Should we constraint the `opens` only to `server` package?

Reply:
> Why we didn't have a similar opens statement for security extension? But it still works?

Do you mean the qualified export (`exports org.elasticsearch.plugins.interceptor to org.elasticsearch.security`) for RestInterceptorActionPlugin ? If so, IIUC that one does not use reflections and the qualified export just limits who can see the interface to implement it. 

> Should we constraint the opens only to server package?

YES ! ... but after hours (literally!) I could not figure out why it would not work. I think I tried every combination and threw every known module into the `to` , but everytime I tried to qualify BOTH the export and opens it failed with error [1] that didn't make any sense (I am confident that error text is wrong). I could qualify either the opens or the exports, but not both. I don't know if this is not supported or if I am bumping into weird/buggy behavior. I could not find any bug reports from java or gradle for this. I think it is fine since the qualified export limits compile time exposure to the interface. It would be nice to lock down the reflective access as well ... and maybe it is. It is very difficult to find documentation on exactly how qualified opens and exports work together. 

[1]
```java
java.lang.IllegalAccessError: superinterface check failed: class co.elastic.elasticsearch.stateless.license.ServerlessLicense (in module org.elasticsearch.stateless) cannot access class org.elasticsearch.license.internal.MutableLicenseService (in module org.elasticsearch.xcore) because module org.elasticsearch.xcore does not export org.elasticsearch.license.internal to module org.elasticsearch.stateless
  at java.base/java.lang.ClassLoader.defineClass1(Native Method)
  at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1013)
  at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1091)
  at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:182)
  at java.base/jdk.internal.loader.Loader.defineClass(Loader.java:612)
  at java.base/jdk.internal.loader.Loader.lambda$findClassInModuleOrNull$6(Loader.java:589)
  at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
  at java.base/jdk.internal.loader.Loader.findClassInModuleOrNull(Loader.java:590)
  at java.base/jdk.internal.loader.Loader.loadClass(Loader.java:548)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
  at java.base/java.lang.Class.forName0(Native Method)
  at java.base/java.lang.Class.forName(Class.java:496)
  at java.base/java.lang.Class.forName(Class.java:475)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.spi.SPIClassIterator.next(SPIClassIterator.java:161)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.createExtensions(PluginsService.java:353)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService$1.loadExtensions(PluginsService.java:340)
  at org.elasticsearch.xcore@8.8.0-SNAPSHOT/org.elasticsearch.xpack.core.XPackPlugin.loadExtensions(XPackPlugin.java:488)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadExtensionsForPlugin(PluginsService.java:346)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadExtensions(PluginsService.java:304)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadBundles(PluginsService.java:293)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.<init>(PluginsService.java:165)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.lambda$getPluginsServiceCtor$16(PluginsService.java:666)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.node.Node.<init>(Node.java:419)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.node.Node.<init>(Node.java:326)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch$2.<init>(Elasticsearch.java:216)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:216)
  at org.elasticsearch.server@8.8.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:67)
```

Reply:
> Do you mean the qualified export (exports org.elasticsearch.plugins.interceptor to org.elasticsearch.security) for RestInterceptorActionPlugin ? If so, IIUC that one does not use reflections and the qualified export just limits who can see the interface to implement it.

No, I mean `org.elasticsearch.xpack.core.security.SecurityExtension` which is also a SPI. The security package module file has 
`exports org.elasticsearch.xpack.core.security;`
but not any `opens` statement. So my question is why do we need `opens` statement here while `SecurityExtension` does not need it?

I also agree we can seek answers to these questions outside of this PR.

Reply:
> So my question is why do we need opens statement here while SecurityExtension does not need it?

That's a good question and I don't have answer. However the [specification](https://docs.oracle.com/javase/specs/jls/se11/html/jls-7.html#jls-7.7) says the following:
 
> The exports directive specifies the name of a package to be exported by the current module. For code in other modules, this grants access at compile time and run time to the public and protected types in the package, and the public and protected members of those types ([§6.6](https://docs.oracle.com/javase/specs/jls/se11/html/jls-6.html#jls-6.6)). It also grants reflective access to those types and members for code in other modules.

> The opens directive specifies the name of a package to be opened by the current module. For code in other modules, this grants access at run time, but not compile time, to the public and protected types in the package, and the public and protected members of those types. It also grants reflective access to all types in the package, and all their members, for code in other modules. 

https://www.logicbig.com/tutorials/core-java-tutorial/modules/reflective-access.html helps to extend that the differences...so I think I have been asking the wrong question (how to qualify opens) and instead should be trying to answer your question "why do we need opens statement here while SecurityExtension does not need it?"   I don't see why we should need opens at all, but it simply does not work without it. I think the answer may have something to do with Guice especially since this case the interface is bound in Guice but in the `SecurityExtension` it is not....but that is just a not-well-educated guess.